set.c

```diff
@@ -0,0 +1,2149 @@
+ *  returns self.
+ */
+static VALUE
+set_i_merge(int argc, VALUE *argv, VALUE set)
+{
+    if (rb_keyword_given_p()) {
+        rb_raise(rb_eArgError, "no keywords accepted");
+    }
+    rb_check_frozen(set);
+
+    int i;
+
+    for (i=0; i < argc; i++) {
+        set_merge_enum_into(set, argv[i]);
+    }
+
+    return set;
+}
+
+static VALUE
+set_reset_table_with_type(VALUE set, const struct st_hash_type *type)
+{
+    rb_check_frozen(set);
+
+    struct set_object *sobj;
+    TypedData_Get_Struct(set, struct set_object, &set_data_type, sobj);
+    set_table *old = &sobj->table;
+
+    size_t size = set_table_size(old);
+    if (size > 0) {
+        set_table *new = set_init_table_with_size(NULL, type, size);
+        struct set_merge_args args = {
+            .set = set,
+            .into = new
+        };
+        set_iter(set, set_merge_i, (st_data_t)&args);
+        set_free_embedded(sobj);
+        memcpy(&sobj->table, new, sizeof(*new));
+        free(new);
+    }
+    else {
+        sobj->table.type = type;
+    }
+
+    return set;
+}
+
+/*
+ *  call-seq:
+ *    compare_by_identity -> self
+ *
+ *  Makes the set compare its elements by their identity and returns self.
+ */
+static VALUE
+set_i_compare_by_identity(VALUE set)
+{
+    if (RSET_COMPARE_BY_IDENTITY(set)) return set;
+
+    if (set_iterating_p(set)) {
+        rb_raise(rb_eRuntimeError, "compare_by_identity during iteration");
+    }
+
+    return set_reset_table_with_type(set, &identhash);
+}
+
+/*
+ *  call-seq:
+ *    compare_by_identity? -> true or false
+ *
+ *  Returns true if the set will compare its elements by their
+ *  identity.  Also see Set#compare_by_identity.
+ */
+static VALUE
+set_i_compare_by_identity_p(VALUE set)
+{
+    return RBOOL(RSET_COMPARE_BY_IDENTITY(set));
+}
+
+/*
+ *  call-seq:
+ *    size -> integer
+ *
+ *  Returns the number of elements.
+ */
+static VALUE
+set_i_size(VALUE set)
+{
+    return RSET_SIZE_NUM(set);
+}
+
+/*
+ *  call-seq:
+ *    empty? -> true or false
+ *
+ *  Returns true if the set contains no elements.
+ */
+static VALUE
+set_i_empty(VALUE set)
+{
+    return RBOOL(RSET_EMPTY(set));
+}
+
+static int
+set_xor_i(st_data_t key, st_data_t data)
+{
+    VALUE set = (VALUE)data;
+    set_table *table = RSET_TABLE(set);
+    key = set_stringify(table, (VALUE)key);
+    RB_OBJ_WRITTEN(set, Qundef, (VALUE)key);
+    if (set_insert(table, key)) {
+        set_delete(table, &key);
+    }
+    return ST_CONTINUE;
+}
+
+/*
+ *  call-seq:
+ *    set ^ enum -> new_set
+ *
+ *  Returns a new set containing elements exclusive between the set and the
+ *  given enumerable object.  <tt>(set ^ enum)</tt> is equivalent to
+ *  <tt>((set | enum) - (set & enum))</tt>.
+ *
+ *    Set[1, 2] ^ Set[2, 3]                   #=> #<Set: {3, 1}>
+ *    Set[1, 'b', 'c'] ^ ['b', 'd']           #=> #<Set: {"d", 1, "c"}>
+ */
+static VALUE
+set_i_xor(VALUE set, VALUE other)
+{
+    other = rb_funcall(rb_obj_class(set), id_new, 1, other);
```

Comment:
There is a is a few places like this where we call `Set.new` from C. Should we? Or should we just call `set_alloc` and sometime benefit from knowing the resulting size upfront?

I suppose it changes the behavior in case `Set.new` is replaced / decorated.

Reply:
It does change behavior in that case, which is why I did it originally.  But I think it makes sense to switch to direct allocation, similar to how we call C functions instead of Ruby methods in general inside the implementation.